### PR TITLE
CodeReview of 13 from 5

### PR DESCRIPTION
**Pull request by lucamoroz**
Dear Mr. Pratesi, 
I hope this message finds you well. 
Please find below the files containing the code that I would kindly ask you to review. 
The methods involved in this review are inside PhotoTester class:
   - PhotoTester
   - testAndReport
   - writeReportToExternalStorage
   - class RunnableTest
   - addTestElement
Please be aware that I've also added two files that don't need any kind of revision, as for now class TestElement is tested by class TestElementTest (using jUnit), but you should be aware of the available methods.

### CODE REVIEW

File: app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        When all the tasks are completed, it saves a report in JSON format and return it as String.
+         */
+        final JSONObject fullJsonReport = new JSONObject();
+
+        int totalTestElements = testElements.size();
+        //countDownLatch allows to sync this thread with the end of all the single tests
+        CountDownLatch countDownLatch = new CountDownLatch(totalTestElements);
+
+        int max_concurrent_tasks = Runtime.getRuntime().availableProcessors();
+        //Define a thread executor that will run a maximum of 'max_concurrent_tasks' simultaneously
```
Comment (leopra) [inline app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java:156]: Why did you need to create simultaneous threads?

File: app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+            }
+        }
+    }
+
+
+    /**
+     * Elaborate tests using threads, stores a report in JSON format to testReport.txt inside the directory given on construction
+     * @return String in JSON format with the test's report, each object is a single test named with the filename and contains:
+     * ingredients, tags, notes, original photo name, recognized text, confidence and alterations (if any), each alteration
+     * contains tags, notes, recognized text, confidence
```
Comment (leopra) [inline app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java:138]: I did not really understand what is "confidence", i think it could be more explanatory

**Review by leopra — COMMENTED** [review]
Overall the code is well done, i appreciate the quantity of comments.

File: app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java (hunk 1 above)
Comment (lucamoroz) [reply app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java:156]: Because this method will potentially run more that 200 tests, and doing this tests simultaneously using the available cores can reduce the time required

File: app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java (hunk 2 above)
Comment (lucamoroz) [reply app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java:138]: Thank you for reporting that. Confidence is a float that describes the correspondence between the recognized text from the OCR and the real one.
